### DISCUSSION TITLE
New rule to detect potential bug with the use of Java stream allMatch

### DISCUSSION THREAD

**Discussion by cah-trent-brown** [pmd/pmd#4372] [Ideas]
When using allMatch() on an empty stream, the method will return true.  This is non-intuitive behavior and can result in bugs if the developer doesn't check that the stream or collection is empty before doing the allMatch().  It would be nice to have a PMD rule that could warn the developer that they have used allMatch() with checking for an empty collection/stream.

**Comment by oowekyala** [comment]
Hi and thanks for the suggestion. I personally don't think we should have a rule for this in the main pmd distribution. This behavior of allMatch is consistent with the behavior of other stream methods (anyMatch, noneMatch), which also have a corner case when the stream is empty. That just needs to be learned with the API in my opinion. Also there is no way to check a stream for emptiness, and then calling allMatch, because the stream would be consumed by the first terminal operation.

Actually, I would recommend a rule that does the opposite of your suggestion and detects redundant emptiness checks on collections before allMatch/anyMatch calls. Such a rule would help teach this corner case behavior to developers unfamiliar with the api.
